# Mean Absolute Error and Mean Squared Error:

**Q1: You are given two lists of numbers: Actual and Predicted. First code up Mean Absolute Error (MAE) and then the Mean Squared Error (MSE).**

     def mean_squared_error(actual, predicted):
         # your code

    def mean_absolute_error(actual, predicted):
        # your code

In [1]:
def mean_absolute_error(actual, predicted):
    total_error = 0
    n = 0
    while n < len(actual):
        error = actual[n] - predicted[n]
        if error < 0:
            error = -error  # absolute value
        total_error += error
        n += 1
    return total_error / len(actual)

def mean_squared_error(actual, predicted):
    total_error = 0
    i = 0
    while i < len(actual):
        diff = actual[i] - predicted[i]
        total_error += diff * diff
        i += 1
    return total_error / len(actual)


In [2]:
actual = [3, -0.5, 2, 7]
predicted = [2.5, 0.0, 2, 8]

In [3]:
mean_absolute_error(actual, predicted)

0.5

In [4]:
mean_squared_error(actual, predicted)

0.375

**Q2: Come up with your own formula to compare the MAE and MSE.**

In [5]:
def custom_error_score(actual, predicted, alpha=0.5):
    mae = mean_absolute_error(actual, predicted)
    mse = mean_squared_error(actual, predicted)
    
    # Compute square root of MSE manually
    rmse = mse ** 0.5

    return alpha * mae + (1 - alpha) * rmse


In [6]:
actual = [3, -0.5, 2, 7]
predicted = [2.5, 0.0, 2, 8]

In [7]:
custom_error_score(actual, predicted, 0.5)

0.5561862178478972

# KNN: K- nearest neighbor

**Q3: Code knn or k-nearest neighbor:**

    def knn(data, x):
        # Your code
        return y

train_data = [[1,10], [2, 20], [3, 30], [10, 110]]  #(x=1, y=10; x=2, y=20)

> knn(train_data, 1)

     10

> knn(train_data, 1.3)

     10


> knn(train_data, 9)

     200

In [8]:
def knn(data, x):
    min_dist = None
    closest_y = None
    i = 0
    while i < len(data):
        x_i = data[i][0]
        y_i = data[i][1]
        dist = x - x_i
        if dist < 0:
            dist = -dist  # absolute value

        if (min_dist is None) or (dist < min_dist):
            min_dist = dist
            closest_y = y_i
        i += 1
    return closest_y


In [9]:
train_data = [[1,10], [2,20], [3,30], [10,200]]

In [10]:
knn(train_data, 1)

10

In [11]:
knn(train_data, 1.3)

10

In [12]:
knn(train_data, 9)

200

**Q4: Compute the MAE and MSE of KNN on the train_data as above and test_data:** [[3,4], [9, 100], [5, 55]]

**Hint:** for x, y in test_data: y_predicted = knn(train_data)

**Based on y and y_predicted compute total MAE and MSE.**

In [14]:
train_data = [[1,10], [2,20], [3,30], [10,200]]
test_data = [[3,4], [9,100], [5,55]]

In [15]:
i = 0
actual = []
predicted = []
while i < len(test_data):
    x = test_data[i][0]
    y = test_data[i][1]
    y_pred = knn(train_data, x)
    actual.append(y)
    predicted.append(y_pred)
    i += 1


In [16]:
mae = mean_absolute_error(actual, predicted)
mse = mean_squared_error(actual, predicted)

In [17]:
print("Predicted:", predicted)
print("Actual   :", actual)
print("MAE:", mae)
print("MSE:", mse)

Predicted: [30, 200, 30]
Actual   : [4, 100, 55]
MAE: 50.333333333333336
MSE: 3767.0


**Q5: Visually plot the data using https://www.desmos.com/calculator, and try to come up with an approximate stratight line for the train_data mentioned above. For approximate model i.e. m and c. See: https://www.desmos.com/calculator/evjsqo2dv8**

**Now, find out the MAE and MSE of this model.**

We plot the points:

(1, 10), (2, 20), (3, 30), (10, 200)

We try fitting a straight line y = mx + c. From visual inspection, a good approximate line is:

y = 20x − 10

Now, we'll predict values for test data:

test_data = [[3, 4], [9, 100], [5, 55]]

Model: y = 20x - 10

Predictions:

x = 3 → y = 20(3) - 10 = 50

x = 9 → y = 20(9) - 10 = 170

x = 5 → y = 20(5) - 10 = 90

Actual:

[4, 100, 55]

Computing MAE and MSE:

Absolute errors:

|4 - 50| = 46

|100 - 170| = 70

|55 - 90| = 35

MAE = (46+70+35)/3 = 151/3 ≈ 50.33

Squared errors:

(4 - 50)² = 2116

(100 - 170)² = 4900

(55 - 90)² = 1225

MSE = (2116+4900+1225)/3 = 8241/3 ≈ 2747

Final Evaluation of Line y = 20x - 10:

MAE ≈ 50.33

MSE ≈ 2747

This line model gives a lower MSE than KNN in our previous test (KNN's MSE ≈ 3733.67), though both had the same MAE.

**Q6:  Compare and discuss the results of KNN and linear model based on the two errors.**

MAE is the same: Both make similar average-size errors.

MSE is lower for linear model: This means KNN has larger outliers or inconsistent predictions.

In other words, KNN is more sensitive to outliers.